File: 0189 rotate-array/solution.go
func rotate(nums []int, k int)  {
    k=k%len(nums)
    rev(nums,0,len(nums)-1)
    rev(nums,0,k-1)
    rev(nums,k,len(nums)-1)
}

func rev(nums []int,i int,j int) {
    for i<j {
        nums[i],nums[j]=nums[j],nums[i]
        i++
        j--
    }
}

File: 0189 rotate-array/solution.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function rotate(nums: number[], k: number): void {
    k=k%nums.length;
    rev(nums,0,nums.length-1);
    rev(nums,0,k-1);
    rev(nums,k,nums.length-1);
};

function rev(nums,l,r) {
    while(l<=r) {
        let temp = nums[l];
        nums[l++] = nums[r];
        nums[r--] = temp;
    }
}
